src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,119 @@
+import slimeknights.tconstruct.library.tools.layout.LayoutSlot;
+import slimeknights.tconstruct.library.tools.part.IToolPart;
+import slimeknights.tconstruct.tools.TinkerTools;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ToolBuildingCategory implements IRecipeCategory<ToolBuildingRecipe> {
+  private static final ResourceLocation BACKGROUND_LOC = TConstruct.getResource("textures/gui/jei/tinker_station.png");
+  private static final Component TITLE = TConstruct.makeTranslation("jei", "tinkering.tool_building");
+  @Getter
+  private final IDrawable icon;
+  @Getter
+  private final IDrawable background;
+  private final IDrawable anvil, slot;
+  private static final int WIDTH = 134;
+  private static final int HEIGHT = 66;
+  private static final int X_OFFSET = -6;
+  private static final int Y_OFFSET = -15;
+  private static final int SLOT_SIZE = 18;
+  private static final int ITEM_SIZE = 16;
+
+  public ToolBuildingCategory(IGuiHelper guiHelper) {
+    this.icon = guiHelper.createDrawableItemStack(TinkerTools.pickaxe.get().getRenderTool());
+    this.background = guiHelper.createDrawable(BACKGROUND_LOC, 122, 77, WIDTH, HEIGHT);
+    this.slot = guiHelper.createDrawable(BACKGROUND_LOC, 24, 14, SLOT_SIZE, SLOT_SIZE);
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<IToolPart> parts = recipe.getToolParts();
+    List<List<ItemStack>> allParts = recipe.getAllToolParts();
+    List<List<ItemStack>> extras = recipe.getExtraRequirements().stream().map(ingredient -> Arrays.asList(ingredient.getItems())).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    if (parts.size() + extras.size() > layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count is greater than layout slot count for %s!", recipe.getId()));
+      int additionalSlots = 0;
+      layoutSlots = new ArrayList<>(layoutSlots);
+      while (parts.size() + extras.size() > layoutSlots.size()) {
+        layoutSlots.add(new LayoutSlot(null, null, additionalSlots * SLOT_SIZE, 0, null));
+        additionalSlots++;
+      }
+    }
+
+    if (parts.size() + extras.size() < layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count is less than layout slot count for %s!", recipe.getId()));
+      extras = new ArrayList<>(extras);
+      while (parts.size() + extras.size() < layoutSlots.size()) {
+        extras.add(List.of(Items.BARRIER.getDefaultInstance()));
+      }
+    }
+
+    for (int i = 0; i < layoutSlots.size(); i++) {
+      builder.addSlot(RecipeIngredientRole.INPUT, layoutSlots.get(i).getX() + X_OFFSET, layoutSlots.get(i).getY() + Y_OFFSET)
+             .addItemStacks(allParts.get(i < parts.size() ? i : i - parts.size()));
```

Comment:
this should be using `extras` when the indexes get higher, not whatever you are doing with that indexing